supabase/migrations/30_manual_billing.sql

```diff
@@ -0,0 +1,354 @@
+    'task_usage_hours', o_task_hours,
+    'trial_credit', coalesce(o_trial_credit, 0),
+    'trial_start', o_trial_start
+  );
+
+end
+$$ language plpgsql volatile security definer;
+
+-- Note: have to redefine this to know about internal.billing_report
+-- Calculate the specified month's billing report for every tenant
+-- and save those reports to billing_historicals.
+create or replace function internal.freeze_billing_month(billed_month timestamptz)
+returns integer as $$
+declare
+    tenant_row record;
+    tenant_count integer = 0;
+begin
+    for tenant_row in select tenant as tenant_name from tenants loop
+        insert into billing_historicals
+        select
+            report->>'billed_prefix' as tenant,
+            (report->>'billed_month')::timestamptz as billed_month,
+            report
+        from internal.billing_report_202308(tenant_row.tenant_name, billed_month) as report
+        on conflict do nothing;
+
+        -- INSERT statements set FOUND true if at least one row is affected, false if no row is affected.
+        if found then
+          tenant_count = tenant_count + 1;
+        end if;
+    end loop;
+    return tenant_count;
+end
+$$ language plpgsql volatile;
+-- End internal billing report gen
+
+create or replace view invoices_ext as
+with has_bypassrls as (
+  select exists(select 1 from pg_roles where rolname = current_role and rolbypassrls = true) as bypass
+),
+authorized_tenants as (
+  select tenants.tenant
+  from tenants
+  left join has_bypassrls on true
+  left join auth_roles('admin') on tenants.tenant ^@ auth_roles.role_prefix
+  where has_bypassrls.bypass or auth_roles.role_prefix is not null
+),
+historical_bills as (
+  select
+    date_trunc('month', (report->>'billed_month')::timestamptz) as date_start,
+    date_trunc('month', (report->>'billed_month')::timestamptz) + interval '1 month' - interval '1 day' as date_end,
+    report->>'billed_prefix' as billed_prefix,
+    report->'line_items' as line_items,
+    report->'subtotal' as subtotal,
+    report as extra
+  from billing_historicals
+  -- inner join should give only rows that match the join condition
+  inner join authorized_tenants on billing_historicals.tenant ^@ authorized_tenants.tenant
+),
+manual_bills as (
+  select
+    date_start,
+    date_end,
+    manual_bills.tenant as billed_prefix,
+    jsonb_build_array(
+      jsonb_build_object(
+        'description', manual_bills.description,
+        'count', 1,
+        'rate', manual_bills.usd_cents,
+        'subtotal', manual_bills.usd_cents
+      )
+    ) as line_items,
+    to_jsonb(usd_cents) as subtotal,
+    'null'::jsonb as extra
+  from manual_bills
+  inner join authorized_tenants on manual_bills.tenant ^@ authorized_tenants.tenant
+),
+current_month as (
+  select
+    date_trunc('month', (report->>'billed_month')::timestamptz) as date_start,
+    date_trunc('month', (report->>'billed_month')::timestamptz) + interval '1 month' - interval '1 day' as date_end,
```

Comment:
this is pedantic, but technically these `date-time`'s don't cover the whole month -- the last date is missing. This stands out to me in the test fixtures below.

It's a little unclear if these response date-end's are inclusive or exclusive. That's true of `manual_bills` as well -- is date_end included in the billing period, or not?

Particularly as these are rendered as full date-times in the API response, my knee-jerk inclination is that they should be an [inclusive, exclusive) range like `2023-07-01T00:00:00Z` to `2023-08-01T00:00:00Z`.

Reply:
This is a good thing to be specific about. My intent was for all dates to only end up being represented as `YYYY-mm-dd` with every date being inclusive. I used the `date` type for `manual_bills.date_start|date_end`, but incorrectly sent back `timestamptz`-formatted values. I _think_ the right thing to do here is change `invoices_ext` to return `date_start`/`date_end` as `YYYY-mm-dd`. That way when we send `"date_start": "2023-08-01", "date_end": "2023-08-31"` it'll be clear that that should cover the entire month. Thoughts on this?